### PR TITLE
KAFKA-14569: Migrate Connect's integration test EmbeddedKafkaCluster from ZK to KRaft mode

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-14569)

This is a squash and rebase of https://github.com/apache/kafka/pull/13375, which is originally authored by @yashmayya. Some light follow-up commits are added to cover changes introduced between now and when the source PR was last updated, and to try to address testing flakiness that seems exacerbated by the switch to KRaft.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@yashmayya apologies for the delay in review for https://github.com/apache/kafka/pull/13375. I've squashed and rebased the PR, ~and verified that all Connect and MirrorMaker 2 tests pass locally~. If you have time, would you be willing to review?

EDIT: Oops, I missed some test cases, TODOs, and possibly even build failures for MM2 (😱). Downgrading to a draft until I've corrected everything.

**Comment by C0urante** [comment]
Okay, this should be good to go--verified the build locally and ran the SSL MM2 integration tests; hoping that's enough coverage for now, and that Jenkins will confirm the rest.

**Comment by C0urante** [comment]
Spoke too soon, but should be good to go now. The one current failing MM2 test appears to be due to the bug that's fixed by https://github.com/apache/kafka/pull/16598, so it should be safe to ignore.

**Comment by C0urante** [comment]
Found one more flaky test. It doesn't seem like that flakiness is inherent with KRaft mode, but it may be exacerbated by it. Pushed a fix.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:121]: Have you tried having a separate controller? My concern is that we have a method `stopOnlyBrokers()` which would stop controllers too in combined mode.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: This was already there before but to be pedantic we'd need `:SSL` in the protocol map to be sure SSL is enabled. This would return `true` for `NOT_SSL:PLAINTEXT`.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Do we need that? This is the default value

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExactlyOnceSourceIntegrationTest.java
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExactlyOnceSourceIntegrationTest.java]: Could we use `VALIDATION_DURATION_MS` here?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR! It looks great overall. I just left a few suggestions

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:121]: Ah, interesting! I think it's possible to use a separate controller node/cluster but it's unnecessary for the purposes of our tests. IMO we should instead just rename the `stopOnlyBrokers` and `restartOnlyBrokers` methods to `stop` and `restart`, respectively.

For the Connect integration tests, we largely treat Kafka like a black box that's usually on and, for some scenarios, may be turned off (or generally unavailable). FWICT we don't really do anything that requires the kind of granularity where we'd distinguish between brokers being unavailable and controllers being unavailable.

It's also a little tricky to use separate controllers and brokers because we still have to be able to turn our embedded Kafka cluster off and back on again with the same port, which becomes difficult when you have to specify a fixed port to use across restarts for brokers only. You end up having to abuse the `setPerServerProperties` API to accomplish that, and while it's certainly possible, it is a bit ugly to expose that cleanly in our internal `EmbeddedKafkaCluster` API.

Thoughts?

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: I <3 pedantry, thanks for the catch! Updated to include `:` and to also recognize `SASL_SSL` as an SSL-enabled listener protocol.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Correct, we can remove. I think it was left over from a lateral refactor that moved from a custom `putIfAbsent` implementation to using the Java standard library.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExactlyOnceSourceIntegrationTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExactlyOnceSourceIntegrationTest.java]: They're the same value, but we're retrying for slightly different reasons; delays in ACL propagation won't necessarily be caught during pre-flight connector config validation.

I've pulled this out into a separate `ACL_PROPAGATION_TIMEOUT_MS` value; LMKWYT.

**Comment by C0urante** [comment]
Thanks Mickael! Pushed a new commit, ready for another round when you have a moment.

**Comment by C0urante** [comment]
Well, I tried to push a new commit at least. Looks like airplane wifi is strong enough to publish GitHub comments but not enough to push commits. Will keep retrying until the commit lands 🥴

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: Is it possible to leverage `ClusterInstance` instead of using `KafkaClusterTestKit` directly? We had introduced the new test infra to some storage tests (https://github.com/apache/kafka/blob/defcbb51ee98ae766e4fcf41ed7631ef1b0cdd15/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerRestartTest.java#L50)

Leveraging the `ClusterInstance` can have following benefits:

1. unify the test style
2. easy to test for all types (zk, kraft, co-kraft)
3. easy to migrate

**Review by chia7712 — COMMENTED** [review]
@C0urante thanks for this nice patch. I have a major question for the migration.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: Or we can migrate the tests having few uses of `EmbeddedKafkaCluster` to `ClusterInstance`. For instance: `OffsetsApiIntegrationTest`. The other tests sticking on `EmbeddedKafkaCluster` can be migrated later.

Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: I file https://issues.apache.org/jira/browse/KAFKA-17174 to address above comments. Please feel free to raise objection. thanks!

Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: For another benefit: #16499 introduces thread-leak detection for method level, and it is applied automatically if we adopt `ClusterTestExtensions` :smile:

Comment (mimaison) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: Thanks for the heads up! There are so many refactorings in progress, it's hard to keep up.

Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: I just want to apply the new test infra to all modules. That can unify the code base and all modules can derive the benefits when we improve it.

BTW, I don't want to block this PR, but it seems to me some tests in connect module can leverage the new test infra first. We can work on them concurrently :)

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: I agree that a unified testing style will be better in the long run, but we're pretty far away from that with Connect's integration tests.

For the vast majority of our tests, we can't easily switch over to the `ClusterTestExtensions` API because our `EmbeddedConnectCluster` automatically spins up its own `EmbeddedKafkaCluster`. If we used the annotation-based API, we'd have to modify our `EmbeddedConnectCluster` class to not create its own embedded Kafka cluster and to instead accept a `ClusterInstance` that's been created and possibly started out-of-band, which would actually be less convenient than our current embedded cluster API and IMO is not a path we should pursue.

I also don't think that adopting the `ClusterTestExtensions` API piecemeal in Connect (like the proposal in KAFKA-17174 to migrate the `OffsetsApiIntegrationTest`) without a long-term strategy to migrate our entire collection of integration tests is a good idea. Right now Connect uses a consistent style for our all of our integration tests and fragmenting that style will just make life harder for the people actually writing and maintaining these tests.

TL;DR: While I'm also not opposed to adopting the `ClusterTestExtensions` API in Connect, I don't think we should start work on that until we have a plan in place for how we can do it for all of our integration tests instead of just some.

**Comment by C0urante** [comment]
@mimaison It looks like my latest commits went through. Would you mind giving this another round when you have a chance? Happy to continue the conversation about the `ClusterTestExtensions` API too, but it seems that that can take place concurrently with a review of the latest changes.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:119]: @C0urante Thanks for your response. I move the discussion to https://issues.apache.org/jira/browse/KAFKA-17174 to avoid unnecessary noise in this PR :)

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: not sure why we need to catch exception? It seems the state is just a inner variable. Maybe this catch is out-of-date.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:322]: Is this a kind of fail fast? otherwise, the admin will get the same error from server.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java
Comment (chia7712) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:1430]: `NUM_PARTITIONS` should be replaced by `numPartitions`. or we can remove `numPartitions` directly

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: If we want to close controllers, we have to call shutdown on controllers explicitly. In combine mode, broker and controller just share the `SharedServer`. `SharedServer` is NOT closed if one of broker/controller is still using. Also, controller instance is not closed when broker instance gets shutdown. That means we are able to connect to controllers even though `stopTemporarily` is called.

The kafka is a black box to connect, so maybe revising the comments is good enough. The purpose of this method is to make kafka unavailable. Also, I guess there is no tests which are used to verify controller connection.

**Review by chia7712 — COMMENTED** [review]
@C0urante thanks for this nice migration. There are some comments using the word "zk".

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java#L83

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java#L96

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java#L123

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectStandalone.java#L50

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Sure, removed the try-catch 👍

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:322]: This logic is basically unchanged from the current state of things; the only reason it shows up in the diff is because there's a slightly different internal API for counting the number of brokers. I'm not sure why we had it originally, but unless it would leave to a regression I'd prefer not to block this PR on figuring that out and save it for a follow-up. Is that alright?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java
Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:1430]: Thanks, good catch! Removed `numPartitions`.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Ah, thanks for the clarification! I made this change in response to [a prior comment](https://github.com/apache/kafka/pull/16599#discussion_r1682557520) that had me thinking that running in combined mode meant it would be impossible to stop brokers separately from controllers. Given this new information I've reverted the commit I made in response to that comment.

I think it should be enough (at least for the sake of this migration) to just run in combined mode. We could definitely add tests later on to verify controller connection.

**Comment by C0urante** [comment]
Thanks @chia7712 for the review! Pushed a few commits that should hopefully address your comments.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:322]: > unless it would leave to a regression I'd prefer not to block this PR on figuring that out and save it for a follow-up. Is that alright?

make senses to me :)

**Comment by chia7712** [comment]
The failure will be fixed by #16688

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
@C0urante Could you please rebase code to include fix #16688 for one more CI?

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (OmniaGM) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Do we care about num of controller == num of brokers can't be just hardcoded to 1. Bit concerned about cases where some one set unsuitable number for `numBrokers` not knowing this will impact quorum as well

Comment (chia7712) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: If the quorum controllers are black box to connect, maybe we can hard code the controllers to `3` (or `1` to reduce the resources in testing)

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: Great call! I tested locally and fixing the number of controller nodes at 1 seemed to reduce the runtime of the `InternalTopicsIntegrationTest::testCreateInternalTopicsWithDefaultSettings` case by about 6-7 seconds on my laptop. There aren't many other tests that use multiple brokers but I've hardcoded the number of controllers to 1 in my latest commit just to play it safe.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java
Comment (mimaison) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java]: That will teach me to trust method names! Thanks for the explanation

**Comment by C0urante** [comment]
I've noticed that the `IdentityReplicationIntegrationTest::testReplication` case has failed four times on this PR branch with the same cause:
```
org.opentest4j.AssertionFailedError: topic config was not synced ==> expected: <compact> but was: <delete>
	at app//org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testReplication(MirrorConnectorsIntegrationBaseTest.java:308)
	at java.base@21.0.3/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base@21.0.3/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base@21.0.3/java.util.ArrayList.forEach(ArrayList.java:1596)
```
AFAIK this exact failure hasn't been occurring on trunk. I'd like to investigate a little more before merging.

**Comment by C0urante** [comment]
Think I found the root cause. Pushed a fix, will try to get three CI runs without a failure on that test case before merging.

**Review by chia7712 — APPROVED** [review]
Those failed tests pass on my local, so LGTM again

**Comment by C0urante** [comment]
Thanks @chia7712!

**Pull request merged**
